lib/bosh/BoshDirectorClient.js

```diff
@@ -16,27 +16,133 @@ const NotFound = errors.NotFound;
+    this.infrastructure = this.config.infrastructure;
+    this.populateCache();
+  }
+
+  clearCache() {
+    this.cache = {};
+  }
+
+  static getActiveConfigs() {
+    return _.filter(config.directors, function (director) {
+      return director.primary;
+    });
+  }
+
+  static getCoreConfigs(activeConfigs) {
+    return _.filter(activeConfigs, function (director) {
+      return director.supportCreate;
+    });
+  }
+
+  static getInactiveConfigs() {
+    return _.filter(config.directors, function (director) {
+      return !director.primary;
     });
-    this.uuid = boshDirector.uuid;
-    this.cpi = boshDirector.cpi;
+  }
+
+  static getPrimaryConfig(coreConfigs) {
+    return _.sample(coreConfigs);
+  }
+
+  getConfigByName(name) {
+    return _.head(_.filter(config.directors, (director) => director.name === name));
+  }
+
+  populateCache() {
+    this.clearCache();
+    return this
+      .getDeployments()
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  addInactiveToCache() {
+    return this
+      .getDeployments(true)
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  deleteCacheEntry(deploymentName) {
+    return delete this.cache[deploymentName];
+  }
+
+  getDirectorConfig(deploymentName) {
+    return Promise.try(() => {
+      logger.info(`Finding the correct director config for:`, deploymentName);
+      if (deploymentName === undefined) {
+        return this.config;
+      }
+      const cache_val = this.cache[deploymentName];
+      if (cache_val !== undefined) {
+        logger.debug('found director in cache...', cache_val);
+        return cache_val;
+      }
+      logger.debug('cache miss for', deploymentName);
+      return this
+        .addInactiveToCache()
+        .then(() => {
+          const cache_val = this.cache[deploymentName];
+          if (cache_val !== undefined) {
+            return cache_val;
+          } else {
+            throw new errors.NotFound(`Deployment not found in directors`, deploymentName);
+          }
+        });
+    });
+  }
+
+  makeRequest(requestDetails, expectedStatusCode, deploymentName) {
+    return this.getDirectorConfig(deploymentName)
+      .then(directorConfig => {
+        requestDetails.baseUrl = directorConfig.url;
+        requestDetails.auth = {
+          user: directorConfig.username,
+          pass: directorConfig.password
+        };
+        requestDetails.rejectUnauthorized = !directorConfig.skip_ssl_validation;
+        return this.request(requestDetails, expectedStatusCode);
+      });
+  }
+
+  makeRequestWithConfig(requestDetails, expectedStatusCode, directorConfig) {
+    requestDetails.baseUrl = `${directorConfig.url}`;
+    requestDetails.auth = {
+      user: directorConfig.username,
+      pass: directorConfig.password
+    };
+    return this.request(requestDetails, expectedStatusCode);
   }
 
   getInfo() {
     return this
-      .request({
+      .makeRequest({
```

Comment:
this method should return an array of director configs now. all directors within 'activePrimary' configuration should be returned back and not just the one activePrimary.